### PR TITLE
Added support for arbitrary tensors for FRN.

### PR DESCRIPTION
**Pull request by Opletts**
I'll write the test cases, just wanted to confirm if my changes are correct.
@Squadrick @saurabhme
In case of FC Layers with dimensions of B x N, we normalize across N. What's the shape of `self.gamma` and `self.beta` supposed to be? I've used 1 x N for now, let me know if I've misunderstood something.

### CODE REVIEW

**Comment by Opletts** [comment]
@googlebot  I signed it!

File: tensorflow_addons/layers/normalizations.py [hunk 1]
```diff
@@ -455,36 +452,36 @@ def get_config(self):
...
+        ndims = len(input_shape)
         for idx, x in enumerate(self.axis):
             if x < 0:
                 self.axis[idx] = ndims + x
 
-        # Validate axes
-        for x in self.axis:
-            if x < 0 or x >= ndims:
-                raise ValueError("Invalid axis: %d" % x)
+        if isinstance(self.axis, list):
```
[the remark this answers is not in the document]
Comment (Squadrick) [reply tensorflow_addons/layers/normalizations.py]: A better way to tackle this is to convert `int` to a list of size 1 in `_check_axis`. 

```python
if isinstance(axis, list): 
    self.axis = axis
elif isinstance(axs, int): 
    self.axis = [axis]
else: 
    raise TypeError(...)
```

File: tensorflow_addons/layers/normalizations.py [hunk 2]
```diff
@@ -455,36 +452,36 @@ def get_config(self):
...
+                elif x == 0 or x == -len(input_shape):
+                    raise ValueError(
+                        "You are trying to normalize your batch axis. "
+                        "Use tf.layers.batch_normalization instead."
+                    )
 
-        if len(self.axis) != len(set(self.axis)):
-            raise ValueError("Duplicate axis: %s" % self.axis)
+        else:
+            self.axis = self.channel_idx = -1
```
Comment (Squadrick) [inline tensorflow_addons/layers/normalizations.py]: Try not to change the user's params. For 2D tensors, we want the user to explicitly use `self.axis = -1` and `self.channel_idx = -1` (default). If they pass any other value, we should be raising an error instead.

File: tensorflow_addons/layers/normalizations.py [hunk 3]
```diff
@@ -350,10 +351,8 @@ class FilterResponseNormalization(tf.keras.layers.Layer):
     Input shape
         Arbitrary. Use the keyword argument `input_shape`
         (tuple of integers, does not include the samples axis)
-        when using this layer as the first layer in a model. This layer, as of now,
-        works on a 4-D tensor where the tensor should have the shape [N X H X W X C]
-
-        TODO: Add support for NCHW data format and FC layers.
+        when using this layer as the first layer in a model. This layer supports
+        arbitrary tensors.
```
Comment (Squadrick) [inline tensorflow_addons/layers/normalizations.py]: Mention the assumptions made: 
1. Excepts the input tensor to be at least 2D.
2. Excepts the batch dim to be `0` in all cases.

File: tensorflow_addons/layers/normalizations.py [hunk 4]
```diff
@@ -497,9 +494,9 @@ def _check_if_input_shape_is_none(self, input_shape):
 
     def _add_gamma_weight(self, input_shape):
         # Get the channel dimension
-        dim = input_shape[-1]
-        shape = [1, 1, 1, dim]
-        # Initialize gamma with shape (1, 1, 1, C)
+        dim = input_shape[self.channel_idx]
+        shape = [1 if i != self.channel_idx else dim for i in range(len(input_shape))]
```
[the remark this answers is not in the document]
Comment (Opletts) [reply tensorflow_addons/layers/normalizations.py:498]: Just thought LC would be quicker, will revert to this.

File: tensorflow_addons/layers/normalizations.py [hunk 5]
```diff
@@ -497,9 +494,9 @@ def _check_if_input_shape_is_none(self, input_shape):
...
+        # Initialize gamma with ones except the channel axis
```
[the remark this answers is not in the document]
Comment (Opletts) [reply tensorflow_addons/layers/normalizations.py:499]: Oops, that's what I meant to write.

File: tensorflow_addons/layers/normalizations.py [hunk 6]
```diff
@@ -455,36 +452,36 @@ def get_config(self):
...
             )
         else:
             self.axis = axis
 
-        if self.axis != [1, 2]:
-            raise ValueError(
-                """FilterResponseNormalization operates on per-channel basis.
-                Axis values should be a list of spatial dimensions."""
-            )
-
```
[the remark this answers is not in the document]
Comment (Opletts) [reply tensorflow_addons/layers/normalizations.py:487]: Shouldn't `len(self.axis)` be restricted to 2 since you want axis values of spatial dimensions? Or can they normalize over more than 2 dimensions?

File: tensorflow_addons/layers/normalizations.py [hunk 7]
```diff
@@ -455,36 +452,36 @@ def get_config(self):
...
-        # Validate axes
-        for x in self.axis:
-            if x < 0 or x >= ndims:
-                raise ValueError("Invalid axis: %d" % x)
+        if isinstance(self.axis, list):
+            for x in self.axis:
+                if x < 0 or x >= len(input_shape):
+                    raise ValueError("Invalid axis: %d" % x)
+
+                elif x == 0 or x == -len(input_shape):
```
[the remark this answers is not in the document]
Comment (Opletts) [reply tensorflow_addons/layers/normalizations.py:465]: Yep, got it.

File: tensorflow_addons/layers/normalizations.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Opletts) [reply tensorflow_addons/layers/normalizations.py]: Adding back `len(self.axis) != len(set(self.axis))` takes care of the size issue.
If any value is greater than `len(input_shape)`, an error will be raised in `_create_input_spec`.

Currently, the user will be able to normalize over all the axes (other than `0`) if they wish to. Does that need to be changed?

File: tensorflow_addons/layers/normalizations.py [hunk 8]
```diff
@@ -455,35 +454,42 @@ def get_config(self):
...
+
+        elif isinstance(axis, int):
+            if axis != -1 or self.channel_idx != -1:
+                raise ValueError("Expected index for 2D is -1 but got {}".format(axis))
+
+            self.axis = [axis]
+
+        else:
             raise TypeError(
                 """Expected a list of values but got {}.""".format(type(axis))
```
Comment (Squadrick) [inline tensorflow_addons/layers/normalizations.py:488]: Expected a *list of values or int*...

Comment (Opletts) [reply tensorflow_addons/layers/normalizations.py:488]: Will do, also do you require type hint for axis to be changed to `Union[list, int]`?

Comment (Squadrick) [reply tensorflow_addons/layers/normalizations.py:488]: Yup.

File: tensorflow_addons/layers/normalizations.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Squadrick) [reply tensorflow_addons/layers/normalizations.py]: All axes other than `0` and `channel_idx`.

**Comment by Opletts** [comment]
@Squadrick Modified the test cases, limited the tensor sizes to 7 because my system ran out of memomry, shoudn't be an issue though.

File: tensorflow_addons/layers/tests/normalizations_test.py [hunk 9]
```diff
@@ -432,14 +539,36 @@ def test_serialization(dtype):
...
 @pytest.mark.usefixtures("maybe_run_functions_eagerly")
 @pytest.mark.parametrize("dtype", [np.float16, np.float32, np.float64])
-def test_eps_gards(dtype):
+def test_eps_grads(dtype):
     set_random_seed()
-    random_inputs = np.random.rand(10, 32, 32, 3).astype(np.float32)
-    random_labels = np.random.randint(2, size=(10,)).astype(np.float32)
-    input_layer = tf.keras.layers.Input(shape=(32, 32, 3))
+
+    shape = np.random.choice(range(1, 30), np.random.randint(2, 7), replace=True)
```
Comment (Squadrick) [inline tensorflow_addons/layers/tests/normalizations_test.py:545]: All tests will run on the same input shape and tensor. This is because `set_random_seeds()` is called at the start of every test function. 

Can you instead test for tensors ranging from 2-D to 5-D, and use randomly pick the `axis`, `channel_idx`?

**Comment by Opletts** [comment]
@Squadrick @AakashKumarNain  just went through the paper again and it doesn't make sense for 2D tensors. FRN is a _global normalisation_ over the spatial extent and that it's done on a _per channel basis_.

- For 4D tensors with `shape = [B x H x W x C]`, it makes sense that we normalise over the spatial dimensions `[H, W]`.
- For 3D tensors, it **HAS** to be of form `[B x H x C]` or `[B x C x H]`, otherwise FRN doesn't make sense.
- 2D tensors have to be of form `[H x C]` and if it's a single sample, the user should reshape it to `[1 x H x C]` and then use FRN.

Is my understanding completely flawed? Was I doing the right thing earlier?

**Comment by Squadrick** [comment]
@Opletts You're right. We assume that the 0th index is batch, so we need 3D at least.

**Comment by Opletts** [comment]
@Squadrick @AakashKumarNain let me know if everything looks alright, I'll make changes if required.

**Comment by seanpmorgan** [comment]
Thank you for your contribution. We sincerely apologize for any delay in reviewing, but TensorFlow Addons is transitioning to a minimal maintenance and release mode. New features will not be added to this repository. For more information, please see our public messaging on this decision:
[TensorFlow Addons Wind Down](https://github.com/tensorflow/addons/issues/2807)

Please consider sending feature requests / contributions to other repositories in the TF community with a similar charters to TFA:
[Keras](https://github.com/keras-team/keras)
[Keras-CV](https://github.com/keras-team/keras-cv)
[Keras-NLP](https://github.com/keras-team/keras-nlp)
